lib/livebook/runtime/evaluator/tracer.ex

```diff
@@ -0,0 +1,129 @@
+defmodule Livebook.Runtime.Evaluator.Tracer do
+  @moduledoc false
+
+  # Compilation tracer used by the evaluator.
+  #
+  # Events are pre-processed and sent to the group leader, where the
+  # tracer state is accumulated. After evaluation the evaluator reads
+  # the accumulated state.
+
+  alias Livebook.Runtime.Evaluator
+
+  defstruct modules_used: MapSet.new(),
+            modules_defined: %{},
+            aliases_used: MapSet.new(),
+            aliases_defined: %{},
+            requires_used: MapSet.new(),
+            requires_defined: MapSet.new(),
+            imports_used?: false,
+            imports_defined?: false,
+            undefined_vars: MapSet.new()
+
+  @doc false
+  def trace(event, env) do
+    case to_updates(event, env) do
+      [] ->
+        :ok
+
+      updates ->
+        io_proxy = Process.group_leader()
+        Evaluator.IOProxy.tracer_updates(io_proxy, updates)
+    end
+
+    :ok
+  end
+
+  defp to_updates(event, env) do
+    # Note that import/require/alias/defmodule don't trigger `:alias_reference`
+    # for the used alias, so we add it explicitly
+
+    case event do
+      {:import, _meta, module, _opts} when env.module == nil ->
+        [{:module_used, module}, {:alias_used, module}, :import_defined]
+
+      {:imported_function, meta, module, name, _arity} ->
+        var? = Keyword.has_key?(meta, :if_undefined)
+        [{:module_used, module}, {:import_used, name, var?}]
+
+      {:imported_macro, meta, module, name, _arity} ->
+        var? = Keyword.has_key?(meta, :if_undefined)
+        [{:module_used, module}, {:import_used, name, var?}]
+
+      {:alias, _meta, alias, as, _opts} when env.module == nil ->
```

Comment:
Why do you check for `env.module == nil`? Code inside a module can still affect compilation... maybe you meant `env.function == nil`?

Reply:
Ah, good catch! This check was meant only for `{import,require,alias}_defined` because those are not accessible outside the module scope, but we should still introduce dependency via `{import,require,alias}_used`.